### PR TITLE
Tests/test setup

### PR DESCRIPTION
**Pull request by jkolyakov**
# Description

Added jest to backend for testing. Restructured parsing functions to adhere to the single responsibility principle and allow for clearer testing. Added a new function to check whether a question answer pair is valid. 

Added test suites for 3/5 functions in parse.js. All work with mongodb, deleting any entries after running.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added test suites for new functions
- [x] Made sure to test success, failure and edge cases in each test suite

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: mern/server/server.js [hunk 1]
```diff
@@ -1,9 +1,9 @@
-import express from 'express';
-import cors from 'cors';
-import mongoose from 'mongoose'
-import dotenv from 'dotenv'
+import express from "express";
+import cors from "cors";
+import mongoose from "mongoose";
+import dotenv from "dotenv";
```
Comment (EmmaHkw) [inline mern/server/server.js:5]: Good catch!

File: mern/server/server.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
```
Comment (EmmaHkw) [inline mern/server/server.js:4]: good catch!

File: mern/server/utils/parse.js [hunk 3]
```diff
@@ -1,42 +1,72 @@
...
-export default function parse(transcript){
-
-    const lst = []
-    for (var i = 0; i<transcript.length; i++){
-
-        lst.push(createQAs(transcript[i]))
-    }
-
-    return lst
+//Parses the transcript by looking through each 'list' of questions and creating question answer pairs
```
Comment (EmmaHkw) [inline mern/server/utils/parse.js:3]: Great comments that explain what the code is doing in a concise manner!

File: mern/server/tests/parse.test.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  const singleItemQuestions = [
+    { question: "hi", id: "1" },
+    { question: "bye", id: "2" },
+    { question: "hello", id: "3" },
+    { question: "goodbye", id: "4" },
+    { question: "alo", id: "5" },
+  ];
+  const noQuestions = [];
+
+  test("Empty list of questions", () => {
```
Comment (EmmaHkw) [inline mern/server/tests/parse.test.js:35]: Great tests as you included corner cases!

File: mern/server/tests/parse.test.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  test("Incorrect JSON object", () => {
+    expect(isValidItem(invalidQuestion) == false);
+  });
+
+  test("Duplicate id", () => {
+    expect(isValidItem(validQuestion) == false);
+  });
+
+  test("Valid question, unique ID", () => {
+    expect(isValidItem(validQuestion2) == true);
```
Comment (EmmaHkw) [inline mern/server/tests/parse.test.js:113]: this item is valid to be added to mongo!

File: mern/server/models/question_answer.js [hunk 6]
```diff
@@ -1,27 +1,29 @@
...
+import mongoose from "mongoose";
 
 const Schema = mongoose.Schema;
 
 const QASchema = new Schema({
-    id: { 
-        type: String,
-        require: true,
-        unique: true
+  id: {
```
Comment (EmmaHkw) [inline mern/server/models/question_answer.js:6]: Are you changing the format here or are they still the same? (I can't seem to find the difference)😅

File: mern/server/server.js [hunk 7]
```diff
@@ -1,9 +1,9 @@
-import express from 'express';
-import cors from 'cors';
-import mongoose from 'mongoose'
-import dotenv from 'dotenv'
```
Comment (EmmaHkw) [inline mern/server/server.js:4]: Just to clarify, both single and double quotations work right?

File: mern/server/utils/parse.js [hunk 8]
```diff
@@ -1,42 +1,72 @@
...
-    //qlist is a list of JSON objects that exists as questions/answer pairs
-    for (var i = 0; i <qlist.length; i++){
-
-        createQA(qlist[i])
-    }
-
-    // Return the first item in the list which should be the top level question
-    return initial
+//Loops through all questions in the JSON object qlist, making them into objects in the databse
+export function createQAs(qlist) {
```
Comment (EmmaHkw) [inline mern/server/utils/parse.js:15]: It might be better (in terms of SOLID Principles) to split these functions into separate files

File: mern/server/utils/parse.js [hunk 9]
```diff
@@ -1,42 +1,72 @@
...
+  const question_included = item.question_included;
+  const intents = item.intents;
+  if (
+    typeof id === "undefined" ||
+    typeof question === "undefined" ||
+    typeof question_included === "undefined" ||
+    typeof intents === "undefined" ||
+    typeof intents.value === "undefined" ||
+    typeof intents.included === "undefined" ||
+    typeof intents.children === "undefined" ||
```
Comment (EmmaHkw) [inline mern/server/utils/parse.js:66]: If an item does not have children (but has all the other information) should it still be considered invalid? Or are we doing this because we are making our own dummy data thus we can assume all items includes all elements?

File: mern/server/package.json [hunk 10]
```diff
@@ -5,21 +5,44 @@
   "main": "index.js",
   "type": "module",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest --coverage --detectOpenHandles",
```
Comment (EmmaHkw) [inline mern/server/package.json:8]: Does --coverage output the percentage of test coverage and whether there are OpenHandles? Where can we find the information output?

File: mern/server/models/question_answer.js (hunk 6 above)
Comment (EmmaHkw) [inline mern/server/models/question_answer.js:6]: Are you changing the format here or are they still the same? (I can't seem to find the difference)😅

File: mern/server/server.js (hunk 7 above)
Comment (EmmaHkw) [inline mern/server/server.js:4]: Just to clarify, both single and double quotations work in js right?

File: mern/server/tests/parse.test.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
+import {
```
Comment (EmmaHkw) [inline mern/server/tests/parse.test.js:1]: When we have the functions in parse.js separated into different files, it might be better to have these tests separated as well (according to their respective functions).😊

**Review by EmmaHkw — COMMENTED** [review]
Looks really good! I have added some comments throughout the code. The main thing is just splitting the functions in parse.js (to adhere to the SOLID Principle and Clean Architecture), but overall great job!

File: mern/server/server.js (hunk 7 above)
Comment (jkolyakov) [reply mern/server/server.js:4]: Yes, I just have prettier set up so it replaces single quotes with double, so when I save it automatically changes it.

File: mern/server/models/question_answer.js (hunk 6 above)
Comment (jkolyakov) [reply mern/server/models/question_answer.js:6]: They're still the same here. It's just changing the formatting of the file so it's more readable.

File: mern/server/utils/parse.js (hunk 8 above)
Comment (jkolyakov) [reply mern/server/utils/parse.js:15]: True. Need to think about the best way to split these up

File: mern/server/models/question_answer.js (hunk 6 above)
Comment (EmmaHkw) [reply mern/server/models/question_answer.js:6]: Got it thank you!

File: mern/server/server.js (hunk 7 above)
Comment (EmmaHkw) [reply mern/server/server.js:4]: Oh that's super useful!

**Review by TerryTian21 — APPROVED** [review]
LGTM
